[Bug] Plugin won't start on 1.20.4

**Describe the bug**
The plugin isn't working on 1.20.4

**To Reproduce**
1. Simply start the server and wait

**Information to investigate**
- Server Version
1.20.4
- Console logs
[09:53:08 ERROR]: [OfflinePlayers] Error initializing plugin 'OfflinePlayers-refresh-optimize-3.3.1.jar' in folder 'plugins' (Is it up to date?)
java.lang.NoClassDefFoundError: com/palmergames/bukkit/towny/exceptions/TownyException
        at de.snap20lp.offlineplayers.OfflinePlayers.onLoad(OfflinePlayers.java:57) ~[OfflinePlayers-refresh-optimize-3.3.1.jar:?]
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:59) ~[paper-1.20.4.jar:git-Paper-485]
        at io.papermc.paper.plugin.storage.ServerPluginProviderStorage.processProvided(ServerPluginProviderStorage.java:18) ~[paper-1.20.4.jar:git-Paper-485]
        at io.papermc.paper.plugin.storage.SimpleProviderStorage.enter(SimpleProviderStorage.java:39) ~[paper-1.20.4.jar:git-Paper-485]
        at io.papermc.paper.plugin.entrypoint.LaunchEntryPointHandler.enter(LaunchEntryPointHandler.java:36) ~[paper-1.20.4.jar:git-Paper-485]
        at org.bukkit.craftbukkit.v1_20_R3.CraftServer.loadPlugins(CraftServer.java:507) ~[paper-1.20.4.jar:git-Paper-485]
        at net.minecraft.server.dedicated.DedicatedServer.initServer(DedicatedServer.java:274) ~[paper-1.20.4.jar:git-Paper-485]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1131) ~[paper-1.20.4.jar:git-Paper-485]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:319) ~[paper-1.20.4.jar:git-Paper-485]
        at java.lang.Thread.run(Thread.java:1623) ~[?:?]
Caused by: java.lang.ClassNotFoundException: com.palmergames.bukkit.towny.exceptions.TownyException
        at org.bukkit.plugin.java.PluginClassLoader.loadClass0(PluginClassLoader.java:197) ~[paper-api-1.20.4-R0.1-SNAPSHOT.jar:?]
        at org.bukkit.plugin.java.PluginClassLoader.loadClass(PluginClassLoader.java:164) ~[paper-api-1.20.4-R0.1-SNAPSHOT.jar:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:521) ~[?:?]
        ... 10 more
- Spigot or Bukkit
Paper-spigot

--- a/.gitignore
+++ b/.gitignore
@@ -15,3 +15,5 @@ gradle/wrapper/gradle-wrapper.properties
 gradlew
 gradlew.bat
 testing_script.sh
+.gitignore
+.idea/

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,0 +1,4 @@
+# v3.3.2
+
+## Bug Fixes
+- Fixed WorldGuard being a true dependency and not a soft dependency.

--- a/build.gradle
+++ b/build.gradle
@@ -3,7 +3,7 @@ plugins {
 }
 
 group 'de.snap20lp.offlineplayers'
-version '3.3.1'
+version '3.3.2'
 
 repositories {
     maven { url = 'https://repo.md-5.net/content/groups/public' }

--- a/build.gradle
+++ b/build.gradle
@@ -3,7 +3,7 @@ plugins {
 }
 
 group 'de.snap20lp.offlineplayers'
-version '3.2.1'
+version '3.3.1'
 
 repositories {
     maven { url = 'https://repo.md-5.net/content/groups/public' }

--- a/build.gradle
+++ b/build.gradle
@@ -44,7 +44,7 @@ dependencies {
     compileOnly 'com.palmergames.bukkit.towny:towny:0.100.1.18'
     compileOnly 'com.onarandombox.multiverseinventories:multiverse-inventories:4.2.5'
     compileOnly 'com.onarandombox.multiversecore:multiverse-core:4.3.12'
-    compileOnly 'com.gitlab.redstonerevive:Departed-Depot:1.1.0-ALPHA'
+    compileOnly 'com.gitlab.redstonerevive:Departed-Depot:1.3.1-ALPHA'
     compileOnly 'com.sk89q.worldguard:worldguard-bukkit:7.0.0'
 }
 

--- a/build.gradle
+++ b/build.gradle
@@ -29,6 +29,7 @@ repositories {
         name "papermc"
         url "https://papermc.io/repo/repository/maven-public/"
     }
+    maven { url "https://maven.enginehub.org/repo/" }
 }
 
 configurations {
@@ -44,6 +45,7 @@ dependencies {
     compileOnly 'com.onarandombox.multiverseinventories:multiverse-inventories:4.2.5'
     compileOnly 'com.onarandombox.multiversecore:multiverse-core:4.3.12'
     compileOnly 'com.gitlab.redstonerevive:Departed-Depot:1.1.0-ALPHA'
+    compileOnly 'com.sk89q.worldguard:worldguard-bukkit:7.0.0'
 }
 
 test {

--- a/build.gradle
+++ b/build.gradle
@@ -43,6 +43,7 @@ dependencies {
     compileOnly 'com.palmergames.bukkit.towny:towny:0.100.1.18'
     compileOnly 'com.onarandombox.multiverseinventories:multiverse-inventories:4.2.5'
     compileOnly 'com.onarandombox.multiversecore:multiverse-core:4.3.12'
+    compileOnly 'com.gitlab.redstonerevive:Departed-Depot:1.1.0-ALPHA'
 }
 
 test {

--- a/build.gradle
+++ b/build.gradle
@@ -3,7 +3,7 @@ plugins {
 }
 
 group 'de.snap20lp.offlineplayers'
-version '3.1.2'
+version '3.2.1'
 
 repositories {
     maven { url = 'https://repo.md-5.net/content/groups/public' }

--- a/src/main/java/de/snap20lp/offlineplayers/CloneManager.java
+++ b/src/main/java/de/snap20lp/offlineplayers/CloneManager.java
@@ -6,6 +6,8 @@ import de.snap20lp.offlineplayers.events.OfflinePlayerHitEvent;
 import de.snap20lp.offlineplayers.events.OfflinePlayerSpawnEvent;
 import me.libraryaddict.disguise.events.UndisguiseEvent;
 import org.bukkit.*;
+import org.bukkit.attribute.Attribute;
+import org.bukkit.attribute.AttributeInstance;
 import org.bukkit.configuration.ConfigurationSection;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.FileConfiguration;
@@ -181,30 +183,35 @@ public class CloneManager implements Listener { // todo: Perhaps refactor events
 
             OfflinePlayerDespawnEvent offlinePlayerDespawnEvent = new OfflinePlayerDespawnEvent(clone);
             Bukkit.getPluginManager().callEvent(offlinePlayerDespawnEvent);
-            playerJoinEvent.getPlayer().teleport(clone.getCloneEntity().getLocation());
-            playerJoinEvent.getPlayer().getActivePotionEffects().forEach(potionEffect -> playerJoinEvent.getPlayer().removePotionEffect(potionEffect.getType()));
-            playerJoinEvent.getPlayer().addPotionEffects(clone.getCloneEntity().getActivePotionEffects());
+            Player player = playerJoinEvent.getPlayer();
+            player.teleport(clone.getCloneEntity().getLocation());
+            player.getActivePotionEffects().forEach(potionEffect -> player.removePotionEffect(potionEffect.getType()));
+            player.addPotionEffects(clone.getCloneEntity().getActivePotionEffects());
 
-            playerJoinEvent.getPlayer().getInventory().setItemInMainHand(clone.getCloneEntity().getEquipment().getItemInMainHand());
-            playerJoinEvent.getPlayer().getInventory().setItemInOffHand(clone.getCloneEntity().getEquipment().getItemInOffHand());
+//            player.getInventory().setItemInMainHand(clone.getCloneEntity().getEquipment().getItemInMainHand());
+            player.getInventory().setItemInOffHand(clone.getCloneEntity().getEquipment().getItemInOffHand());
             if (clone.getCloneEntity().hasPotionEffect(PotionEffectType.SLOW) && clone.isCloneHasAI()) {
-                playerJoinEvent.getPlayer().removePotionEffect(PotionEffectType.SLOW);
+                player.removePotionEffect(PotionEffectType.SLOW);
             }
             if (clone.isDead()) {
                 if (!OfflinePlayers.getInstance().getConfig().getBoolean("OfflinePlayer.cloneKeepItems")) {
-                    if (playerJoinEvent.getPlayer().getEquipment() != null) {
-                        playerJoinEvent.getPlayer().getEquipment().clear();
+                    if (player.getEquipment() != null) {
+                        player.getEquipment().clear();
                     }
                 }
-                playerJoinEvent.getPlayer().setLevel(0);
-                playerJoinEvent.getPlayer().setExp(0.0f);
-                playerJoinEvent.getPlayer().setHealth(0.0d);
+                player.setLevel(0);
+                player.setExp(0.0f);
+                player.setHealth(0.0d);
             } else {
-                playerJoinEvent.getPlayer().setFireTicks(clone.getCloneEntity().getFireTicks());
-                playerJoinEvent.getPlayer().setHealth(clone.getCloneEntity().getHealth());
+                player.setFireTicks(clone.getCloneEntity().getFireTicks());
+                AttributeInstance attributeInstance = player.getAttribute(Attribute.GENERIC_MAX_HEALTH);
+                double maxHealth = 20.0;
+                if (attributeInstance != null)
+                    maxHealth = attributeInstance.getValue();
+                player.setHealth(Math.min(clone.getCloneEntity().getHealth(), maxHealth));
                 clone.getAddedItems().forEach(itemStack -> {
                     if (itemStack != null) {
-                        playerJoinEvent.getPlayer().getInventory().addItem(itemStack);
+                        player.getInventory().addItem(itemStack);
                     }
                 });
             }

--- a/src/main/java/de/snap20lp/offlineplayers/DepartedDepotIntegration.java
+++ b/src/main/java/de/snap20lp/offlineplayers/DepartedDepotIntegration.java
@@ -30,6 +30,7 @@ public class DepartedDepotIntegration implements Listener {
     public DepartedDepotIntegration () {
         if (Bukkit.getPluginManager().isPluginEnabled("DepartedDepot")) {
             departedDepot = (DepartedDepot) Bukkit.getPluginManager().getPlugin("DepartedDepot");
+            OfflinePlayers.getInstance().getLogger().log(Level.INFO, "We have located Departed Depot. OfflinePlayers will spawn graves on death.");
         } else {
             OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "We could not locate DepartedDepot. Clones will not spawn graves. Everything else should be fine.");
             departedDepot = null;

--- a/src/main/java/de/snap20lp/offlineplayers/DepartedDepotIntegration.java
+++ b/src/main/java/de/snap20lp/offlineplayers/DepartedDepotIntegration.java
@@ -1,0 +1,55 @@
+package de.snap20lp.offlineplayers;
+
+import com.gitlab.redstonerevive.departeddepot.DepartedDepot;
+import com.gitlab.redstonerevive.departeddepot.Depot;
+import com.gitlab.redstonerevive.departeddepot.DepotManager;
+import de.snap20lp.offlineplayers.events.OfflinePlayerDeathEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.inventory.ItemStack;
+
+import java.util.List;
+import java.util.logging.Level;
+
+/**
+ * The DepartedDepotIntegration exists to let clones spawn depots when they die.
+ *
+ * @author Sugaku
+ */
+public class DepartedDepotIntegration implements Listener {
+
+    /**
+     * The reference to the DepartedDepot plugin.
+     */
+    private final DepartedDepot departedDepot;
+
+    /**
+     * Attempts to create the DepartedDepotIntegration module.
+     */
+    public DepartedDepotIntegration () {
+        if (Bukkit.getPluginManager().isPluginEnabled("DepartedDepot")) {
+            departedDepot = (DepartedDepot) Bukkit.getPluginManager().getPlugin("DepartedDepot");
+        } else {
+            OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "We could not locate DepartedDepot. Clones will not spawn graves. Everything else should be fine.");
+            departedDepot = null;
+        }
+    }
+
+    /**
+     * Called whenever a clone dies.
+     *
+     * @param event The clone death event.
+     */
+    @EventHandler
+    public void onDeath (OfflinePlayerDeathEvent event) {
+        if (departedDepot == null) return;
+        OfflinePlayer player = event.getOfflinePlayer();
+        List<ItemStack> items = player.getSavedInventoryContents();
+        DepotManager.getInstance().addDepot(new Depot(player.getCloneEntity().getLocation(),
+                player.getOfflinePlayer().getUniqueId(),
+                player.getPlayerExp(),
+                items.toArray(new ItemStack[0])));
+        player.getSavedInventoryContents().clear();
+    }
+}

--- a/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
+++ b/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
@@ -6,17 +6,11 @@ import com.onarandombox.multiverseinventories.share.Sharables;
 import com.palmergames.bukkit.towny.TownyAPI;
 import com.palmergames.bukkit.towny.exceptions.TownyException;
 import com.palmergames.bukkit.towny.object.Town;
-import com.sk89q.worldedit.bukkit.BukkitAdapter;
-import com.sk89q.worldguard.WorldGuard;
-import com.sk89q.worldguard.protection.ApplicableRegionSet;
-import com.sk89q.worldguard.protection.flags.StateFlag;
-import com.sk89q.worldguard.protection.flags.registry.FlagRegistry;
-import com.sk89q.worldguard.protection.regions.RegionContainer;
-import com.sk89q.worldguard.protection.regions.RegionQuery;
+import de.snap20lp.offlineplayers.depends.APIManager;
+import de.snap20lp.offlineplayers.depends.WorldGuardFacade;
 import de.snap20lp.offlineplayers.events.OfflinePlayerSpawnEvent;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
-import org.bukkit.World;
 import org.bukkit.configuration.file.FileConfiguration;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
@@ -41,11 +35,6 @@ public class EventProtector implements Listener {
     private final String destinationWorld;
 
     /**
-     * This is the instance of the custom flag that we need in order to query its value.
-     */
-    public static StateFlag BAN_OFFLINE_PLAYERS;
-
-    /**
      * Creates a new MortalMaker and loads config file values.
      */
     public EventProtector () {
@@ -60,14 +49,10 @@ public class EventProtector implements Listener {
      * @param event The OfflinePlayerSpawnEvent.
      */
     @EventHandler
-    public void onCloneSpawn (OfflinePlayerSpawnEvent event) {
-        Location location = event.getLocation();
-        World w = location.getWorld();
-        if (w == null) return;
-        RegionContainer container = WorldGuard.getInstance().getPlatform().getRegionContainer();
-        RegionQuery query = container.createQuery();
-        ApplicableRegionSet set = query.getApplicableRegions(BukkitAdapter.adapt(location));
-        if (set.testState(null, BAN_OFFLINE_PLAYERS)) {
+    public void onCloneSpawn(OfflinePlayerSpawnEvent event) {
+        WorldGuardFacade wg = APIManager.getInstance().getWorldGuard();
+        if (wg == null) return;
+        if (wg.testFlag(event.getLocation())) {
             Location spawn = null;
             if (isBedEnabled) spawn = event.getOfflinePlayer().getOfflinePlayer().getBedSpawnLocation();
             TownyAPI api = OfflinePlayers.getTownyAPI();
@@ -76,7 +61,8 @@ public class EventProtector implements Listener {
                 if (town != null) {
                     try {
                         spawn = town.getSpawn();
-                    } catch (TownyException ignored) { }
+                    } catch (TownyException ignored) {
+                    }
                 }
             }
             if (spawn == null) spawn = Bukkit.getWorld(destinationWorld).getSpawnLocation();

--- a/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
+++ b/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
@@ -1,0 +1,109 @@
+package de.snap20lp.offlineplayers;
+
+import com.onarandombox.multiverseinventories.MultiverseInventories;
+import com.onarandombox.multiverseinventories.profile.PlayerProfile;
+import com.onarandombox.multiverseinventories.share.Sharables;
+import com.palmergames.bukkit.towny.TownyAPI;
+import com.palmergames.bukkit.towny.exceptions.TownyException;
+import com.palmergames.bukkit.towny.object.Town;
+import com.sk89q.worldedit.bukkit.BukkitAdapter;
+import com.sk89q.worldguard.WorldGuard;
+import com.sk89q.worldguard.protection.ApplicableRegionSet;
+import com.sk89q.worldguard.protection.flags.StateFlag;
+import com.sk89q.worldguard.protection.flags.registry.FlagRegistry;
+import com.sk89q.worldguard.protection.regions.RegionContainer;
+import com.sk89q.worldguard.protection.regions.RegionQuery;
+import de.snap20lp.offlineplayers.events.OfflinePlayerSpawnEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.World;
+import org.bukkit.configuration.file.FileConfiguration;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import org.bukkit.inventory.ItemStack;
+
+/**
+ * The Event Protector module prevents players from accidentally leaving their offline player in a WorldGuard protected
+ * region. Frequently these regions are used for events, hence the name.
+ *
+ * @author Sugaku
+ */
+public class EventProtector implements Listener {
+
+    /**
+     * Whether bed spawns are allowed or not.
+     */
+    private final boolean isBedEnabled;
+
+    /**
+     * The world to teleport players to the spawn of.
+     */
+    private final String destinationWorld;
+
+    /**
+     * Whether WorldGuard is present on the server.
+     */
+    private final boolean worldGuard;
+
+    /**
+     * This is the instance of the custom flag that we need in order to query its value.
+     */
+    private final StateFlag BAN_OFFLINE_PLAYERS;
+
+    /**
+     * Creates a new MortalMaker and loads config file values.
+     */
+    public EventProtector () {
+        FileConfiguration config = OfflinePlayers.getInstance().getConfig();
+        isBedEnabled = config.getBoolean("OfflinePlayer.event-protector.is-bed-enabled", false);
+        destinationWorld = config.getString("OfflinePlayer.event-protector.destination-world", "world");
+        worldGuard = Bukkit.getPluginManager().isPluginEnabled("WorldGuard");
+        if (worldGuard) {
+            FlagRegistry registry = WorldGuard.getInstance().getFlagRegistry();
+            StateFlag flag = new StateFlag("ban-offline-players", false);
+            registry.register(flag);
+            BAN_OFFLINE_PLAYERS = flag;
+        } else BAN_OFFLINE_PLAYERS = null;
+    }
+
+    /**
+     * Called whenever a clone is spawned because a player is logging off.
+     *
+     * @param event The OfflinePlayerSpawnEvent.
+     */
+    @EventHandler
+    public void onCloneSpawn (OfflinePlayerSpawnEvent event) {
+        if (!worldGuard) return;
+        Location location = event.getLocation();
+        World w = location.getWorld();
+        if (w == null) return;
+        RegionContainer container = WorldGuard.getInstance().getPlatform().getRegionContainer();
+        RegionQuery query = container.createQuery();
+        ApplicableRegionSet set = query.getApplicableRegions(BukkitAdapter.adapt(location));
+        if (set.testState(null, BAN_OFFLINE_PLAYERS)) {
+            Location spawn = null;
+            if (isBedEnabled) spawn = event.getOfflinePlayer().getOfflinePlayer().getBedSpawnLocation();
+            TownyAPI api = OfflinePlayers.getTownyAPI();
+            if (api != null && spawn == null) {
+                Town town = api.getTown(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
+                if (town != null) {
+                    try {
+                        spawn = town.getSpawn();
+                    } catch (TownyException ignored) { }
+                }
+            }
+            if (spawn == null) spawn = Bukkit.getWorld(destinationWorld).getSpawnLocation();
+            event.setLocation(spawn);
+            MultiverseInventories multiInvAPI = OfflinePlayers.getMultiverseInventoriesAPI();
+            if (multiInvAPI != null) {
+                PlayerProfile profile = multiInvAPI.getGroupManager().getGroupsForWorld(spawn.getWorld().getName()).get(0).getGroupProfileContainer().getPlayerData(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
+                ItemStack[] inv = profile.get(Sharables.INVENTORY);
+                ItemStack[] armor = profile.get(Sharables.ARMOR);
+                ItemStack offHand = profile.get(Sharables.OFF_HAND);
+                event.setInventory(inv);
+                event.setArmor(armor);
+                event.setOffHand(offHand);
+            }
+        }
+    }
+}

--- a/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
+++ b/src/main/java/de/snap20lp/offlineplayers/EventProtector.java
@@ -41,14 +41,9 @@ public class EventProtector implements Listener {
     private final String destinationWorld;
 
     /**
-     * Whether WorldGuard is present on the server.
-     */
-    private final boolean worldGuard;
-
-    /**
      * This is the instance of the custom flag that we need in order to query its value.
      */
-    private final StateFlag BAN_OFFLINE_PLAYERS;
+    public static StateFlag BAN_OFFLINE_PLAYERS;
 
     /**
      * Creates a new MortalMaker and loads config file values.
@@ -57,13 +52,6 @@ public class EventProtector implements Listener {
         FileConfiguration config = OfflinePlayers.getInstance().getConfig();
         isBedEnabled = config.getBoolean("OfflinePlayer.event-protector.is-bed-enabled", false);
         destinationWorld = config.getString("OfflinePlayer.event-protector.destination-world", "world");
-        worldGuard = Bukkit.getPluginManager().isPluginEnabled("WorldGuard");
-        if (worldGuard) {
-            FlagRegistry registry = WorldGuard.getInstance().getFlagRegistry();
-            StateFlag flag = new StateFlag("ban-offline-players", false);
-            registry.register(flag);
-            BAN_OFFLINE_PLAYERS = flag;
-        } else BAN_OFFLINE_PLAYERS = null;
     }
 
     /**
@@ -73,7 +61,6 @@ public class EventProtector implements Listener {
      */
     @EventHandler
     public void onCloneSpawn (OfflinePlayerSpawnEvent event) {
-        if (!worldGuard) return;
         Location location = event.getLocation();
         World w = location.getWorld();
         if (w == null) return;

--- a/src/main/java/de/snap20lp/offlineplayers/MortalMaker.java
+++ b/src/main/java/de/snap20lp/offlineplayers/MortalMaker.java
@@ -1,18 +1,14 @@
 package de.snap20lp.offlineplayers;
 
-import com.onarandombox.multiverseinventories.MultiverseInventories;
-import com.onarandombox.multiverseinventories.profile.PlayerProfile;
-import com.onarandombox.multiverseinventories.share.Sharables;
-import com.palmergames.bukkit.towny.TownyAPI;
-import com.palmergames.bukkit.towny.exceptions.TownyException;
-import com.palmergames.bukkit.towny.object.Town;
+import de.snap20lp.offlineplayers.depends.APIManager;
+import de.snap20lp.offlineplayers.depends.MultiverseInventoriesFacade;
+import de.snap20lp.offlineplayers.depends.TownyFacade;
 import de.snap20lp.offlineplayers.events.OfflinePlayerSpawnEvent;
 import org.bukkit.Bukkit;
 import org.bukkit.Location;
 import org.bukkit.configuration.file.FileConfiguration;
 import org.bukkit.event.EventHandler;
 import org.bukkit.event.Listener;
-import org.bukkit.inventory.ItemStack;
 
 import java.util.List;
 
@@ -64,27 +60,12 @@ public class MortalMaker implements Listener {
                         spawn = null;
 
             }
-            TownyAPI api = OfflinePlayers.getTownyAPI();
-            if (api != null && spawn == null) {
-                Town town = api.getTown(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
-                if (town != null) {
-                    try {
-                        spawn = town.getSpawn();
-                    } catch (TownyException ignored) { }
-                }
-            }
+            TownyFacade towny = APIManager.getInstance().getTownyFacade();
+            if (towny != null && spawn == null) spawn = towny.grabTownSpawn(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
             if (spawn == null) spawn = Bukkit.getWorld(destinationWorld).getSpawnLocation();
             event.setLocation(spawn);
-            MultiverseInventories multiInvAPI = OfflinePlayers.getMultiverseInventoriesAPI();
-            if (multiInvAPI != null) {
-                PlayerProfile profile = multiInvAPI.getGroupManager().getGroupsForWorld(spawn.getWorld().getName()).get(0).getGroupProfileContainer().getPlayerData(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
-                ItemStack[] inv = profile.get(Sharables.INVENTORY);
-                ItemStack[] armor = profile.get(Sharables.ARMOR);
-                ItemStack offHand = profile.get(Sharables.OFF_HAND);
-                event.setInventory(inv);
-                event.setArmor(armor);
-                event.setOffHand(offHand);
-            }
+            MultiverseInventoriesFacade multiverseInventoriesFacade = APIManager.getInstance().getMultiverseInventoriesFacade();
+            if (multiverseInventoriesFacade != null) multiverseInventoriesFacade.updateInventory(event);
         }
     }
 }

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
@@ -18,10 +18,8 @@ import org.bukkit.potion.PotionEffectType;
 
 import java.math.BigInteger;
 import java.util.ArrayList;
-import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicInteger;
 
 public class OfflinePlayer implements Listener {
 
@@ -183,32 +181,36 @@ public class OfflinePlayer implements Listener {
             int disguiseID = disguisedEntity.getEntity().getEntityId();
             cloneEntity.getNearbyEntities(distance, distance, distance).forEach(entity -> {
                 if (entity.getEntityId() != entityID && entity.getEntityId() != disguiseID) {
-                    if (entity.getType() == EntityType.PLAYER && !cloneEntityId.equals(offlinePlayer.getUniqueId())) isNearby.set(true);
+                    if (entity.getType() == EntityType.PLAYER && !cloneEntityId.equals(offlinePlayer.getUniqueId()))
+                        isNearby.set(true);
                     else if (entity instanceof Mob mob && canProvoke)
-                        if (mob.getTarget() == null && mob.hasLineOfSight(cloneEntity))
-                            mob.setTarget(cloneEntity);
+                        if (mob.getTarget() == null && mob.hasLineOfSight(cloneEntity)) mob.setTarget(cloneEntity);
                 }
             });
 
             if (!isNearby.get()) {
-                OfflinePlayer offlinePlayerClone = new OfflinePlayer(offlinePlayer, currentSeconds, cloneEntity.getLocation(), playerExp, cloneEntity.getHealth(), savedInventoryContents, savedArmorContents,
-                        /*cloneEntity.getEquipment().getItemInMainHand()*/ new ItemStack(Material.AIR, 1), cloneEntity.getEquipment().getItemInOffHand());
-                offlinePlayerClone.replaceCloneStats(cloneEntity);
+//                OfflinePlayer offlinePlayerClone = new OfflinePlayer(offlinePlayer, currentSeconds, cloneEntity.getLocation(), playerExp, cloneEntity.getHealth(), savedInventoryContents, savedArmorContents,
+//                        /*cloneEntity.getEquipment().getItemInMainHand()*/ new ItemStack(Material.AIR, 1), cloneEntity.getEquipment().getItemInOffHand());
+//                offlinePlayerClone.replaceCloneStats(cloneEntity);
 
                 despawnClone();
-                Map<UUID, OfflinePlayer> offlinePlayerList = CloneManager.getInstance().getOfflinePlayerList();
-                Map<Integer, OfflinePlayer> entityList = CloneManager.getInstance().getEntityOfflinePlayerHashMap();
-                offlinePlayerList.remove(offlinePlayer.getUniqueId());
-                entityList.remove(entityID);
-                cancelUpdateTask();
-                cancelDespawnTask();
-
-                offlinePlayerList.put(offlinePlayer.getUniqueId(), offlinePlayerClone);
-                entityList.put(offlinePlayerClone.getDisguisedEntity().getEntity().getEntityId(), offlinePlayerClone);
+//                Map<UUID, OfflinePlayer> offlinePlayerList = CloneManager.getInstance().getOfflinePlayerList();
+//                Map<Integer, OfflinePlayer> entityList = CloneManager.getInstance().getEntityOfflinePlayerHashMap();
+//                offlinePlayerList.remove(offlinePlayer.getUniqueId());
+//                entityList.remove(entityID);
+//                cancelUpdateTask();
+//                cancelDespawnTask();
+//                spawnClone();
+//                despawnClone();
+                isHidden = true;
+//                startTimers();
+//                offlinePlayerList.put(offlinePlayer.getUniqueId(), offlinePlayerClone);
+//                entityList.put(offlinePlayerClone.getDisguisedEntity().getEntity().getEntityId(), offlinePlayerClone);
             } else if (isNearby.get() && isHidden) {
                 spawnClone();
                 isHidden = false;
             }
+
             BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
             runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
             sum = sum.add(time); // todo: Remove used for benchmarking performance.
@@ -329,9 +331,12 @@ public class OfflinePlayer implements Listener {
     public void despawnClone() {
         if (cloneEntity != null) {
             DisguiseAPI.undisguiseToAll(cloneEntity);
+            disguisedEntity.removeDisguise();
             cloneEntity.remove();
-            cancelUpdateTask();
-            cloneEntity.remove();
+            if (cloneEntityId != null) {
+                Entity e = Bukkit.getEntity(cloneEntityId);
+                if (e != null) e.remove();
+            }
         }
     }
 

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
@@ -179,50 +179,41 @@ public class OfflinePlayer implements Listener {
             long startTime = System.nanoTime(); // todo: Remove used for benchmarking performance.
             if (isDead) return;
             AtomicBoolean isNearby = new AtomicBoolean(false);
+            int entityID = cloneEntity.getEntityId();
+            int disguiseID = disguisedEntity.getEntity().getEntityId();
             cloneEntity.getNearbyEntities(distance, distance, distance).forEach(entity -> {
-                if (entity.getEntityId() != cloneEntity.getEntityId() && entity.getEntityId() != disguisedEntity.getEntity().getEntityId()) {
-                    if (entity.getType() == EntityType.PLAYER && !entity.getUniqueId().equals(offlinePlayer.getUniqueId())) {
-                        isNearby.set(true);
-                    } else if (entity instanceof Mob && canProvoke) {
-                        if (((Mob) entity).getTarget() == null && ((Mob) entity).hasLineOfSight(cloneEntity))
-                            ((Mob) entity).setTarget(cloneEntity);
-                    }
+                if (entity.getEntityId() != entityID && entity.getEntityId() != disguiseID) {
+                    if (entity.getType() == EntityType.PLAYER && !cloneEntityId.equals(offlinePlayer.getUniqueId())) isNearby.set(true);
+                    else if (entity instanceof Mob mob && canProvoke)
+                        if (mob.getTarget() == null && mob.hasLineOfSight(cloneEntity))
+                            mob.setTarget(cloneEntity);
                 }
             });
 
             if (!isNearby.get()) {
-
                 OfflinePlayer offlinePlayerClone = new OfflinePlayer(offlinePlayer, currentSeconds, cloneEntity.getLocation(), playerExp, cloneEntity.getHealth(), savedInventoryContents, savedArmorContents,
                         /*cloneEntity.getEquipment().getItemInMainHand()*/ new ItemStack(Material.AIR, 1), cloneEntity.getEquipment().getItemInOffHand());
                 offlinePlayerClone.replaceCloneStats(cloneEntity);
 
-                DisguiseAPI.undisguiseToAll(cloneEntity);
-                cloneEntity.remove();
                 despawnClone();
                 Map<UUID, OfflinePlayer> offlinePlayerList = CloneManager.getInstance().getOfflinePlayerList();
                 Map<Integer, OfflinePlayer> entityList = CloneManager.getInstance().getEntityOfflinePlayerHashMap();
                 offlinePlayerList.remove(offlinePlayer.getUniqueId());
-                if (cloneEntityId != null) {
-                    Entity e = Bukkit.getEntity(cloneEntityId);
-                    if (e != null) e.remove();
-                }
-                entityList.remove(cloneEntity.getEntityId());
+                entityList.remove(entityID);
                 cancelUpdateTask();
                 cancelDespawnTask();
 
                 offlinePlayerList.put(offlinePlayer.getUniqueId(), offlinePlayerClone);
                 entityList.put(offlinePlayerClone.getDisguisedEntity().getEntity().getEntityId(), offlinePlayerClone);
-            }
-
-            if (isNearby.get() && isHidden) {
+            } else if (isNearby.get() && isHidden) {
                 spawnClone();
                 isHidden = false;
             }
             BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
             runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
             sum = sum.add(time); // todo: Remove used for benchmarking performance.
-            System.out.println("Finished: " + time.divide(BigInteger.valueOf(10000)) + "centimilliseconds"); // todo: Remove used for benchmarking performance.
-            System.out.println("Average: " + (sum.divide(runs.multiply(BigInteger.valueOf(10000)))) + "centimilliseconds"); // todo: Remove used for benchmarking performance.
+            System.out.println("Finished: " + time.divide(BigInteger.valueOf(10000)) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
+            System.out.println("Average: " + (sum.divide(runs.multiply(BigInteger.valueOf(10000)))) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
         }, 10, cloneUpdateTimer);
 
         if (despawnTimerEnabled) {
@@ -337,6 +328,8 @@ public class OfflinePlayer implements Listener {
 
     public void despawnClone() {
         if (cloneEntity != null) {
+            DisguiseAPI.undisguiseToAll(cloneEntity);
+            cloneEntity.remove();
             cancelUpdateTask();
             cloneEntity.remove();
         }

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
@@ -171,8 +171,8 @@ public class OfflinePlayer implements Listener {
     }
 
     // todo: Remove used for benchmarking performance.
-    BigInteger sum = BigInteger.ZERO;
-    BigInteger runs = BigInteger.ONE;
+    static BigInteger sum = BigInteger.ZERO;
+    static BigInteger runs = BigInteger.ONE;
 
     public void startTimers() {
         updateTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(OfflinePlayers.getInstance(), () -> {
@@ -220,8 +220,9 @@ public class OfflinePlayer implements Listener {
             }
             BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
             runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
-            System.out.println("Finished: " + time + "ns"); // todo: Remove used for benchmarking performance.
-            System.out.println("Average: " + (sum.divide(runs)) + "ns"); // todo: Remove used for benchmarking performance.
+            sum = sum.add(time); // todo: Remove used for benchmarking performance.
+            System.out.println("Finished: " + time.divide(BigInteger.valueOf(10000)) + "centimilliseconds"); // todo: Remove used for benchmarking performance.
+            System.out.println("Average: " + (sum.divide(runs.multiply(BigInteger.valueOf(10000)))) + "centimilliseconds"); // todo: Remove used for benchmarking performance.
         }, 10, cloneUpdateTimer);
 
         if (despawnTimerEnabled) {

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
@@ -174,7 +174,7 @@ public class OfflinePlayer implements Listener {
 
     public void startTimers() {
         updateTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(OfflinePlayers.getInstance(), () -> {
-            long startTime = System.nanoTime(); // todo: Remove used for benchmarking performance.
+//            long startTime = System.nanoTime(); // todo: Remove used for benchmarking performance.
             if (isDead) return;
             AtomicBoolean isNearby = new AtomicBoolean(false);
             int entityID = cloneEntity.getEntityId();
@@ -211,11 +211,11 @@ public class OfflinePlayer implements Listener {
                 isHidden = false;
             }
 
-            BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
-            runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
-            sum = sum.add(time); // todo: Remove used for benchmarking performance.
-            System.out.println("Finished: " + time.divide(BigInteger.valueOf(10000)) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
-            System.out.println("Average: " + (sum.divide(runs.multiply(BigInteger.valueOf(10000)))) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
+//            BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
+//            runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
+//            sum = sum.add(time); // todo: Remove used for benchmarking performance.
+//            System.out.println("Finished: " + time.divide(BigInteger.valueOf(10000)) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
+//            System.out.println("Average: " + (sum.divide(runs.multiply(BigInteger.valueOf(10000)))) + " centimilliseconds"); // todo: Remove used for benchmarking performance.
         }, 10, cloneUpdateTimer);
 
         if (despawnTimerEnabled) {

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayer.java
@@ -16,10 +16,12 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.potion.PotionEffect;
 import org.bukkit.potion.PotionEffectType;
 
+import java.math.BigInteger;
 import java.util.ArrayList;
 import java.util.Map;
 import java.util.UUID;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
 
 public class OfflinePlayer implements Listener {
 
@@ -49,6 +51,13 @@ public class OfflinePlayer implements Listener {
     private boolean isDead = false;
     private static final boolean despawnTimerEnabled = OfflinePlayers.getInstance().getConfig().getBoolean("OfflinePlayer.cloneDe-spawnTimer.enabled");
     private static final boolean isBlockEntity = OfflinePlayers.getInstance().getConfig().getBoolean("OfflinePlayer.useBlockEntity");
+
+    private static final int distance = OfflinePlayers.getInstance().getConfig().getInt("OfflinePlayer.cloneSpawnDistance");
+
+    private static final boolean canProvoke = OfflinePlayers.getInstance().getConfig().getBoolean("OfflinePlayer.cloneCanProvoke");
+
+    private static final int cloneUpdateTimer = OfflinePlayers.getInstance().getConfig().getInt("OfflinePlayer.cloneUpdateTimer");
+
     private LivingEntity cloneEntity;
     private TargetedDisguise disguisedEntity;
 
@@ -161,16 +170,20 @@ public class OfflinePlayer implements Listener {
         this.spawnLocation = val;
     }
 
+    // todo: Remove used for benchmarking performance.
+    BigInteger sum = BigInteger.ZERO;
+    BigInteger runs = BigInteger.ONE;
+
     public void startTimers() {
         updateTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(OfflinePlayers.getInstance(), () -> {
+            long startTime = System.nanoTime(); // todo: Remove used for benchmarking performance.
             if (isDead) return;
-            int distance = OfflinePlayers.getInstance().getConfig().getInt("OfflinePlayer.cloneSpawnDistance");
             AtomicBoolean isNearby = new AtomicBoolean(false);
             cloneEntity.getNearbyEntities(distance, distance, distance).forEach(entity -> {
                 if (entity.getEntityId() != cloneEntity.getEntityId() && entity.getEntityId() != disguisedEntity.getEntity().getEntityId()) {
                     if (entity.getType() == EntityType.PLAYER && !entity.getUniqueId().equals(offlinePlayer.getUniqueId())) {
                         isNearby.set(true);
-                    } else if (entity instanceof Mob && OfflinePlayers.getInstance().getConfig().getBoolean("OfflinePlayer.cloneCanProvoke")) {
+                    } else if (entity instanceof Mob && canProvoke) {
                         if (((Mob) entity).getTarget() == null && ((Mob) entity).hasLineOfSight(cloneEntity))
                             ((Mob) entity).setTarget(cloneEntity);
                     }
@@ -205,7 +218,11 @@ public class OfflinePlayer implements Listener {
                 spawnClone();
                 isHidden = false;
             }
-        }, 10, OfflinePlayers.getInstance().getConfig().getInt("OfflinePlayer.cloneUpdateTimer"));
+            BigInteger time = BigInteger.valueOf(System.nanoTime() - startTime); // todo: Remove used for benchmarking performance.
+            runs = runs.add(BigInteger.ONE); // todo: Remove used for benchmarking performance.
+            System.out.println("Finished: " + time + "ns"); // todo: Remove used for benchmarking performance.
+            System.out.println("Average: " + (sum.divide(runs)) + "ns"); // todo: Remove used for benchmarking performance.
+        }, 10, cloneUpdateTimer);
 
         if (despawnTimerEnabled) {
             despawnTask = Bukkit.getScheduler().scheduleSyncRepeatingTask(OfflinePlayers.getInstance(), new Runnable() {

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
@@ -54,6 +54,7 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
             multiverseInventoriesAPI = (MultiverseInventories) Bukkit.getPluginManager().getPlugin("Multiverse-Inventories");
         Bukkit.getPluginManager().registerEvents(CloneManager.getInstance(), this);
         Bukkit.getPluginManager().registerEvents(new MortalMaker(), this);
+        Bukkit.getPluginManager().registerEvents(new DepartedDepotIntegration(), this);
         if (getServer().getPluginManager().getPlugin("LibsDisguises") == null || getServer().getPluginManager().getPlugin("ProtocolLib") == null) {
             Bukkit.getConsoleSender().sendMessage("§4[OfflinePlayers] ERROR: LibsDisguises is not activated! Please install LibsDisguises and ProtocolLib to use this plugin!");
             Bukkit.getPluginManager().disablePlugin(this);

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
@@ -2,9 +2,7 @@ package de.snap20lp.offlineplayers;
 
 import com.onarandombox.multiverseinventories.MultiverseInventories;
 import com.palmergames.bukkit.towny.TownyAPI;
-import com.sk89q.worldguard.WorldGuard;
-import com.sk89q.worldguard.protection.flags.StateFlag;
-import com.sk89q.worldguard.protection.flags.registry.FlagRegistry;
+import de.snap20lp.offlineplayers.depends.APIManager;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.block.data.BlockData;
@@ -50,12 +48,7 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
      * Called extra early on in the plugin loading process. Used to register WorldGuard flags.
      */
     public void onLoad () {
-        try {
-            FlagRegistry registry = WorldGuard.getInstance().getFlagRegistry();
-            StateFlag flag = new StateFlag("ban-offline-players", false);
-            registry.register(flag);
-            EventProtector.BAN_OFFLINE_PLAYERS = flag;
-        } catch (Exception ignored) { }
+        APIManager.getInstance().delegatedOnLoad();
     }
 
     @Override

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
@@ -54,6 +54,7 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
             multiverseInventoriesAPI = (MultiverseInventories) Bukkit.getPluginManager().getPlugin("Multiverse-Inventories");
         Bukkit.getPluginManager().registerEvents(CloneManager.getInstance(), this);
         Bukkit.getPluginManager().registerEvents(new MortalMaker(), this);
+        Bukkit.getPluginManager().registerEvents(new EventProtector(), this);
         Bukkit.getPluginManager().registerEvents(new DepartedDepotIntegration(), this);
         if (getServer().getPluginManager().getPlugin("LibsDisguises") == null || getServer().getPluginManager().getPlugin("ProtocolLib") == null) {
             Bukkit.getConsoleSender().sendMessage("§4[OfflinePlayers] ERROR: LibsDisguises is not activated! Please install LibsDisguises and ProtocolLib to use this plugin!");

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
@@ -2,6 +2,9 @@ package de.snap20lp.offlineplayers;
 
 import com.onarandombox.multiverseinventories.MultiverseInventories;
 import com.palmergames.bukkit.towny.TownyAPI;
+import com.sk89q.worldguard.WorldGuard;
+import com.sk89q.worldguard.protection.flags.StateFlag;
+import com.sk89q.worldguard.protection.flags.registry.FlagRegistry;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.block.data.BlockData;
@@ -43,6 +46,18 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
         return multiverseInventoriesAPI;
     }
 
+    /**
+     * Called extra early on in the plugin loading process. Used to register WorldGuard flags.
+     */
+    public void onLoad () {
+        try {
+            FlagRegistry registry = WorldGuard.getInstance().getFlagRegistry();
+            StateFlag flag = new StateFlag("ban-offline-players", false);
+            registry.register(flag);
+            EventProtector.BAN_OFFLINE_PLAYERS = flag;
+        } catch (Exception ignored) { }
+    }
+
     @Override
     public void onEnable() {
         this.metrics = new Metrics(this, 19973);
@@ -54,7 +69,8 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
             multiverseInventoriesAPI = (MultiverseInventories) Bukkit.getPluginManager().getPlugin("Multiverse-Inventories");
         Bukkit.getPluginManager().registerEvents(CloneManager.getInstance(), this);
         Bukkit.getPluginManager().registerEvents(new MortalMaker(), this);
-        Bukkit.getPluginManager().registerEvents(new EventProtector(), this);
+        if (Bukkit.getPluginManager().isPluginEnabled("WorldGuard"))
+            Bukkit.getPluginManager().registerEvents(new EventProtector(), this);
         Bukkit.getPluginManager().registerEvents(new DepartedDepotIntegration(), this);
         if (getServer().getPluginManager().getPlugin("LibsDisguises") == null || getServer().getPluginManager().getPlugin("ProtocolLib") == null) {
             Bukkit.getConsoleSender().sendMessage("§4[OfflinePlayers] ERROR: LibsDisguises is not activated! Please install LibsDisguises and ProtocolLib to use this plugin!");

--- a/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
+++ b/src/main/java/de/snap20lp/offlineplayers/OfflinePlayers.java
@@ -1,7 +1,5 @@
 package de.snap20lp.offlineplayers;
 
-import com.onarandombox.multiverseinventories.MultiverseInventories;
-import com.palmergames.bukkit.towny.TownyAPI;
 import de.snap20lp.offlineplayers.depends.APIManager;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
@@ -17,34 +15,6 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
     }
 
     /**
-     * The active Towny api instance.
-     */
-    private static TownyAPI townyAPI;
-
-    /**
-     * The active Multiverse-Inventories instance.
-     */
-    private static MultiverseInventories multiverseInventoriesAPI;
-
-    /**
-     * Accessor method for the towny api. No promises to not provide null.
-     *
-     * @return The Towny API.
-     */
-    public static TownyAPI getTownyAPI () {
-        return townyAPI;
-    } // todo: Group APIs into an API manager.
-
-    /**
-     * Accessor method for the MultiverseInventories api. No promises to not provide null.
-     *
-     * @return The MultiverseInventories API.
-     */
-    public static MultiverseInventories getMultiverseInventoriesAPI () {
-        return multiverseInventoriesAPI;
-    }
-
-    /**
      * Called extra early on in the plugin loading process. Used to register WorldGuard flags.
      */
     public void onLoad () {
@@ -56,15 +26,11 @@ public class OfflinePlayers extends JavaPlugin { // todo: Maybe create a config 
         this.metrics = new Metrics(this, 19973);
         Bukkit.getConsoleSender().sendMessage("§aOfflinePlayers starting in version " + getDescription().getVersion());
         this.saveDefaultConfig();
-        if (Bukkit.getPluginManager().isPluginEnabled("Towny"))
-            townyAPI = TownyAPI.getInstance();
-        if (Bukkit.getPluginManager().isPluginEnabled("Multiverse-Inventories"))
-            multiverseInventoriesAPI = (MultiverseInventories) Bukkit.getPluginManager().getPlugin("Multiverse-Inventories");
+        APIManager.getInstance().delegatedOnEnabled();
         Bukkit.getPluginManager().registerEvents(CloneManager.getInstance(), this);
         Bukkit.getPluginManager().registerEvents(new MortalMaker(), this);
         if (Bukkit.getPluginManager().isPluginEnabled("WorldGuard"))
             Bukkit.getPluginManager().registerEvents(new EventProtector(), this);
-        Bukkit.getPluginManager().registerEvents(new DepartedDepotIntegration(), this);
         if (getServer().getPluginManager().getPlugin("LibsDisguises") == null || getServer().getPluginManager().getPlugin("ProtocolLib") == null) {
             Bukkit.getConsoleSender().sendMessage("§4[OfflinePlayers] ERROR: LibsDisguises is not activated! Please install LibsDisguises and ProtocolLib to use this plugin!");
             Bukkit.getPluginManager().disablePlugin(this);

--- a/src/main/java/de/snap20lp/offlineplayers/depends/APIManager.java
+++ b/src/main/java/de/snap20lp/offlineplayers/depends/APIManager.java
@@ -1,6 +1,8 @@
 package de.snap20lp.offlineplayers.depends;
 
+import de.snap20lp.offlineplayers.DepartedDepotIntegration;
 import de.snap20lp.offlineplayers.OfflinePlayers;
+import org.bukkit.Bukkit;
 
 import javax.annotation.Nullable;
 import java.util.logging.Level;
@@ -24,6 +26,16 @@ public class APIManager {
     private WorldGuardFacade worldGuard;
 
     /**
+     * The MultiverseInventoriesFacade that is contained in the APIManager.
+     */
+    private MultiverseInventoriesFacade multiverseInventoriesFacade;
+
+    /**
+     * The TownyFacade that is contained in the APIManager.
+     */
+    private TownyFacade townyFacade;
+
+    /**
      * An accessor method to the active APIManager instance.
      *
      * @return The active APIManager instance.
@@ -34,7 +46,7 @@ public class APIManager {
     }
 
     /**
-     * Called when onLoad() is issued executed for OfflinePlayers.
+     * Called when onLoad() is executed for OfflinePlayers.
      */
     public void delegatedOnLoad () {
         try {
@@ -46,6 +58,30 @@ public class APIManager {
     }
 
     /**
+     * Called when onEnable() is executed for OfflinePlayers.
+     */
+    public void delegatedOnEnabled () {
+        try {
+            townyFacade = new TownyFacade();
+        } catch (NoClassDefFoundError error) {
+            townyFacade = null;
+            OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "Towny not found. Plugin will run normally.");
+        }
+        try {
+            multiverseInventoriesFacade = new MultiverseInventoriesFacade();
+        } catch (NoClassDefFoundError error) {
+            multiverseInventoriesFacade = null;
+            OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "Multiverse Inventories not found. Plugin will run normally.");
+        }
+        // todo: Update to match the APIManager pattern.
+        try {
+            Bukkit.getPluginManager().registerEvents(new DepartedDepotIntegration(), OfflinePlayers.getInstance());
+        } catch (NoClassDefFoundError error) {
+            OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "Departed Depots not found. Plugin will run normally.");
+        }
+    }
+
+    /**
      * An accessor method for the WorldGuardFacade.
      *
      * @return The WorldGuardFacade that was created by the APIManager.
@@ -53,5 +89,25 @@ public class APIManager {
     @Nullable
     public WorldGuardFacade getWorldGuard () {
         return worldGuard;
+    }
+
+    /**
+     * An accessor method for the MultiverseInventoriesFacade.
+     *
+     * @return The MultiverseInventoriesFacade that was created by the APIManager.
+     */
+    @Nullable
+    public MultiverseInventoriesFacade getMultiverseInventoriesFacade () {
+        return multiverseInventoriesFacade;
+    }
+
+    /**
+     * An accessor method for the MultiverseInventoriesFacade.
+     *
+     * @return The MultiverseInventoriesFacade that was created by the APIManager.
+     */
+    @Nullable
+    public TownyFacade getTownyFacade () {
+        return townyFacade;
     }
 }

--- a/src/main/java/de/snap20lp/offlineplayers/depends/APIManager.java
+++ b/src/main/java/de/snap20lp/offlineplayers/depends/APIManager.java
@@ -1,0 +1,57 @@
+package de.snap20lp.offlineplayers.depends;
+
+import de.snap20lp.offlineplayers.OfflinePlayers;
+
+import javax.annotation.Nullable;
+import java.util.logging.Level;
+
+/**
+ * The APIManager attempts to isolate all API loading from the main OfflinePlayer's plugin class. These are then further
+ * isolated into individual facades.
+ *
+ * @author Sugaku
+ */
+public class APIManager {
+
+    /**
+     * The active APIManager instance.
+     */
+    private static APIManager instance;
+
+    /**
+     * The WorldGuardFacade that is contained in the APIManager.
+     */
+    private WorldGuardFacade worldGuard;
+
+    /**
+     * An accessor method to the active APIManager instance.
+     *
+     * @return The active APIManager instance.
+     */
+    public static APIManager getInstance () {
+        if (instance == null) instance = new APIManager();
+        return instance;
+    }
+
+    /**
+     * Called when onLoad() is issued executed for OfflinePlayers.
+     */
+    public void delegatedOnLoad () {
+        try {
+            worldGuard = new WorldGuardFacade();
+        } catch (NoClassDefFoundError error) {
+            worldGuard = null;
+            OfflinePlayers.getInstance().getLogger().log(Level.WARNING, "WorldGuard not found. Plugin will run normally.");
+        }
+    }
+
+    /**
+     * An accessor method for the WorldGuardFacade.
+     *
+     * @return The WorldGuardFacade that was created by the APIManager.
+     */
+    @Nullable
+    public WorldGuardFacade getWorldGuard () {
+        return worldGuard;
+    }
+}

--- a/src/main/java/de/snap20lp/offlineplayers/depends/MultiverseInventoriesFacade.java
+++ b/src/main/java/de/snap20lp/offlineplayers/depends/MultiverseInventoriesFacade.java
@@ -1,0 +1,45 @@
+package de.snap20lp.offlineplayers.depends;
+
+import com.onarandombox.multiverseinventories.MultiverseInventories;
+import com.onarandombox.multiverseinventories.profile.PlayerProfile;
+import com.onarandombox.multiverseinventories.share.Sharables;
+import de.snap20lp.offlineplayers.OfflinePlayers;
+import de.snap20lp.offlineplayers.events.OfflinePlayerSpawnEvent;
+import org.bukkit.Bukkit;
+import org.bukkit.inventory.ItemStack;
+
+/**
+ * This serves as a compatibility layer between OfflinePlayers and MultiverseInventories. Creating an instance of this
+ * class without MultiverseInventories will cause errors.
+ *
+ * @author Sugaku
+ */
+public class MultiverseInventoriesFacade {
+
+    /**
+     * The active Multiverse-Inventories instance.
+     */
+    private static MultiverseInventories multiverseInventoriesAPI;
+
+    /**
+     * Creates a new MultiverseInventoriesFacade.
+     */
+    public MultiverseInventoriesFacade () {
+        multiverseInventoriesAPI = (MultiverseInventories) Bukkit.getPluginManager().getPlugin("Multiverse-Inventories");
+    }
+
+    /**
+     * Updates MortalMaker related events using the correct inventories from Multiverse Inventories.
+     *
+     * @param event The OfflinePlayerSpawnEvent to update with the correct inventories.
+     */
+    public void updateInventory (OfflinePlayerSpawnEvent event) {
+        PlayerProfile profile = multiverseInventoriesAPI.getGroupManager().getGroupsForWorld(event.getLocation().getWorld().getName()).get(0).getGroupProfileContainer().getPlayerData(event.getOfflinePlayer().getOfflinePlayer().getPlayer());
+        ItemStack[] inv = profile.get(Sharables.INVENTORY);
+        ItemStack[] armor = profile.get(Sharables.ARMOR);
+        ItemStack offHand = profile.get(Sharables.OFF_HAND);
+        event.setInventory(inv);
+        event.setArmor(armor);
+        event.setOffHand(offHand);
+    }
+}

--- a/src/main/java/de/snap20lp/offlineplayers/depends/TownyFacade.java
+++ b/src/main/java/de/snap20lp/offlineplayers/depends/TownyFacade.java
@@ -1,0 +1,46 @@
+package de.snap20lp.offlineplayers.depends;
+
+import com.palmergames.bukkit.towny.TownyAPI;
+import com.palmergames.bukkit.towny.exceptions.TownyException;
+import com.palmergames.bukkit.towny.object.Town;
+import org.bukkit.Bukkit;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+
+/**
+ * The TownyFacade is used to separate this plugin from specific Towny implementations. Without Towny present, it is
+ * unsafe to instantiate this class.
+ *
+ * @author Sugaku
+ */
+public class TownyFacade {
+
+    /**
+     * The active Towny api instance.
+     */
+    private static TownyAPI townyAPI;
+
+    /**
+     * Creates a new TownyFacade by grabbing the TownyAPI.
+     */
+    public TownyFacade () {
+        townyAPI = (TownyAPI) Bukkit.getPluginManager().getPlugin("Towny");
+    }
+
+    /**
+     * Grabs the town spawn of the given player from Towny.
+     *
+     * @param player The player to grab the town spawn of.
+     */
+    public Location grabTownSpawn (Player player) {
+        Town town = townyAPI.getTown(player);
+        if (town != null) {
+            try {
+                return town.getSpawn();
+            } catch (TownyException ignored) {
+                return null;
+            }
+        }
+        else return null;
+    }
+}

--- a/src/main/java/de/snap20lp/offlineplayers/depends/WorldGuardFacade.java
+++ b/src/main/java/de/snap20lp/offlineplayers/depends/WorldGuardFacade.java
@@ -1,0 +1,49 @@
+package de.snap20lp.offlineplayers.depends;
+
+import com.sk89q.worldedit.bukkit.BukkitAdapter;
+import com.sk89q.worldguard.WorldGuard;
+import com.sk89q.worldguard.protection.ApplicableRegionSet;
+import com.sk89q.worldguard.protection.flags.StateFlag;
+import com.sk89q.worldguard.protection.flags.registry.FlagRegistry;
+import com.sk89q.worldguard.protection.regions.RegionContainer;
+import com.sk89q.worldguard.protection.regions.RegionQuery;
+import org.bukkit.Location;
+import org.bukkit.World;
+
+/**
+ * A Facade sits right between the dependency and the OfflinePlayers plugin. Instantiating one when a dependency is not
+ * found will cause a bug.
+ *
+ * @author Sugaku
+ */
+public class WorldGuardFacade {
+
+    /**
+     * This is the instance of the custom flag that we need in order to query its value.
+     */
+    private final StateFlag BAN_OFFLINE_PLAYERS;
+
+    /**
+     * Attempts to create the WorldGuardFacade. THIS MAY THROW NoClassDefFound!
+     */
+    public WorldGuardFacade () {
+        FlagRegistry registry = WorldGuard.getInstance().getFlagRegistry();
+        StateFlag flag = new StateFlag("ban-offline-players", false);
+        registry.register(flag);
+        BAN_OFFLINE_PLAYERS = flag;
+    }
+
+    /**
+     * Tests the given location with the BAN_OFFLINE_PLAYERS flag.
+     *
+     * @param location The location to test.
+     */
+    public boolean testFlag (Location location) {
+        World w = location.getWorld();
+        if (w == null) return false;
+        RegionContainer container = WorldGuard.getInstance().getPlatform().getRegionContainer();
+        RegionQuery query = container.createQuery();
+        ApplicableRegionSet set = query.getApplicableRegions(BukkitAdapter.adapt(location));
+        return set.testState(null, BAN_OFFLINE_PLAYERS);
+    }
+}

--- a/src/main/resources/config.yml
+++ b/src/main/resources/config.yml
@@ -37,3 +37,7 @@ OfflinePlayer:
     destination-world: 'world'
   death-flavour-messages:
     - " was killed by "
+  event-protector:
+    # Players in protected WorldGuard will be brought to their 'Bed Spawn > Towny Spawn > (Overworld: 'world') World Spawn'
+    is-bed-enabled: true
+    destination-world: 'world'

--- a/src/main/resources/plugin.yml
+++ b/src/main/resources/plugin.yml
@@ -10,3 +10,4 @@ softdepend:
   - LibsDisguises
   - Towny
   - Multiverse-Inventories
+  - DepartedDepot

--- a/src/main/resources/plugin.yml
+++ b/src/main/resources/plugin.yml
@@ -11,3 +11,4 @@ softdepend:
   - Towny
   - Multiverse-Inventories
   - DepartedDepot
+  - WorldGuard

